packages/dds/merge-tree/docs/high-level.md

```diff
@@ -0,0 +1,401 @@
+
+We'll start with the string "abc". We have two clients, "A" and "B". Both clients insert a single character at position 1 at the same time, without knowing about the existence of the other.
+
+Here's what our op stream might look like.
+
+```ts
+[
+    { seq: 1, refSeq: 0, clientId: "A", type: "insert", text: "X", index: 1 },
+    { seq: 2, refSeq: 0, clientId: "B", type: "insert", text: "Y", index: 1 },
+]
+```
+
+Our string would look like this:
+
+start: "abc"  
+seq 1: "aXbc"  
+seq 2: "aYXbc"  
+
+Note that the refSeq of the two ops is the same. This means that the two operations occurred without either client being aware of the other. The sequence number on client A's op is lower than client B's op. This means that client A's op likely reached the server before client B's, though this does not imply anything about the time that either op occurred. It's also worth mentioning that even if the two ops _had_ reached the server in the exact same nanosecond, they would still have different sequence numbers.
+
+Because neither op was aware of the other, we would consider them to be concurrent.
+
+The scenario above looks like this in prose:
+
+When the server processes the op from client A, it sends back to client A the sequence number "1". It then sends to client B the op from client A.
+
+On receiving the sequence number from the server, client A updates its refSeq to be 1.
+
+Before client A's op has been sent to client B, client B sends its own op to the server. Client B receives client A's op and updates its refSeq to 1.
+
+The server does the same thing for client B's op, and both client A and client B update their refSeq to 2. 
+
+This should hopefully start to make a bit of sense, but will become clearer once we talk about the internal representation of the merge-tree and go a bit more in depth on collaboration.
+
+#### Internal Structure
+
+Merge-tree represents a sequence as an ordered series of segments, with individual characters or elements being identified by the combination of a segment index and a character offset into that segment, referred to as "segoff." For example, if we take the string "abcde" and split it into arbitrary segments:
+
+```ts
+["a", "bcd", "e"]
+```
+
+The character "c" in the above string is at segment 1, offset 1. The character "e" is at segment 2, offset 0.
+
+When the endpoints of an operation (like insertion or removal) fall within a segment, that segment is split. For example:
+
+```ts
+["ae"] -> insert "bcd" at position 1 -> ["a", "bcd", "e"]
+["abcde"] -> remove the range [1, 4] -> ["a", "bcd", "e"] -> ["a", "e"]
+```
+
+In order to map a numeric character position to a segoff, the merge-tree traverses the list of segments summing up the length of each segment until it lands inside one. Naively this list of segments can be modeled by a 2d-array, but in practice we represent this as a tree-like structure to efficiently traverse over large sections of the tree at once.
```

Comment:
IMO it's worth drawing the connection between merge-tree and a [rope](https://en.wikipedia.org/wiki/Rope_(data_structure)) here